src/xwayland/xserver.rs

```diff
@@ -1,509 +1,382 @@
- * Xwayland will send SIGUSR1 to the process, and if a thread cannot handle
- * SIGUSR1, that thread will be killed.
- *
- * Double-fork can tackle this issue, but this is also very complex in a
- * a multithread program, after forking only signal-safe functions can be used.
- * The only workaround is to fork early before any other thread starts, but
- * doing so will expose an unsafe interface.
- *
- * We use an intermediate shell to translate the signal to simple fd IO.
- * We ask sh to setup SIGUSR1 handler, and in a subshell mute SIGUSR1 and exec
- * Xwayland. When the SIGUSR1 is received, it can communicate to us via redirected
- * STDOUT.
- */
 use std::{
     env,
     ffi::OsStr,
-    io::{self, Read},
+    os::fd::{BorrowedFd, OwnedFd},
     os::unix::{
-        io::{AsRawFd, BorrowedFd, RawFd},
+        io::{AsRawFd, RawFd},
         net::UnixStream,
         process::CommandExt,
     },
-    process::{ChildStdout, Command, Stdio},
+    process::Command,
     sync::{Arc, Mutex},
 };
 
-use calloop::{
-    channel::{self, sync_channel, Channel, SyncSender},
-    generic::Generic,
-    Interest, LoopHandle, Mode,
-};
-use wayland_server::{
-    backend::{ClientData, ClientId, DisconnectReason},
-    Client, DisplayHandle,
-};
+use tracing::{info, trace};
+use wayland_server::backend::ClientData;
+use wayland_server::backend::DisconnectReason;
+use wayland_server::{Client, DisplayHandle};
 
-use tracing::{error, info, instrument};
+use crate::{utils::user_data::UserDataMap, wayland::compositor::CompositorClientState};
 
 use super::x11_sockets::{prepare_x11_sockets, X11Lock};
-use crate::utils::user_data::UserDataMap;
-#[cfg(feature = "wayland_frontend")]
-use crate::wayland::compositor::CompositorClientState;
 
-/// The XWayland handle
+/// A handle to a running XWayland process. Using XWayland as an xserver for
+/// X11-based clients requires two connections: one wayland socket, where
+/// XWayland creates surfaces for its clients, and one X11 socket, where a
+/// compositor's implementation of an X11 window manager connects and handles
+/// X11 events.
+///            ┌───────────────────┐
+///            │                   │
+///            │                   │ X11 Clients
+///            │     Xwayland      │ ◄─────
+///            │                   │
+///            │                   │
+///            └──┬──────────────▲┘
+///               │              │
+/// Wayland Socket│              │ X11 Socket
+///               │              │
+///               │              │
+///  ┌────────────▼─────────────┼────────────┐
+///  │  Compositor               │            │
+///  │                         ┌─┴───────┐    │
+///  │                         │ XWM     │    │
+///  │                         │         │    │
+///  │                         └─────────┘    │
+///  │                                        │
+///  └────────────────────────────────────────┘
```

Comment:
How does this render in docs? iirc a `text` code block is needed for this type of graphics?

Reply:
Fixed. It's pretty now.